File: examples/33_data_center/process_dipolee_data.py
"""
Process DIPOLee data center power profiles.

Loads raw Excel power data from three sources, computes daily and weekly
percentile bands (0/10/25/50/75/90/100), and plots the results.
"""

import time

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

# ============================================================
# SETUP
# ============================================================
t0 = time.time()

DATA_PATH_SOURCE1 = (
    "/home/cbay/code/data/Samples/NLR/colocation/"
)
DATA_PATH_SOURCE2 = (
    "/home/cbay/code/data/Samples/NLR/inference/"
)
DATA_PATH_SOURCE3 = (
    "/home/cbay/code/data/Samples/ORNL/"
)

SOURCE1_FILES = {
    # "Power1_1MW": "colocation_10MW_2840nodes_20u_power.xlsx",
    # "Power1_2MW": "colocation_10MW_2840nodes_40u_power.xlsx",
    # "Power1_3MW": "colocation_10MW_2840nodes_60u_power.xlsx",
    "Power1_4MW": "colocation_10MW_2840nodes_80u_power.xlsx",
}
SOURCE2_FILES = {
    # "Power2_1MW": "inference_1MW_283nodes_20u_power.xlsx",
    # "Power2_2MW": "inference_1MW_283nodes_40u_power.xlsx",
    # "Power2_3MW": "inference_1MW_283nodes_60u_power.xlsx",
    "Power2_4MW": "inference_1MW_283nodes_80u_power.xlsx",
}
SOURCE3_FILE = "87F42G2C+2QG3-124-131-124-131.xlsx"

# (raw_name, proc_name, tile_id)  tile_id is 1-based, matching MATLAB layout
DATASETS = [
    ("Power1_1MW", "Power1_1", 1),
    ("Power1_2MW", "Power1_2", 4),
    ("Power1_3MW", "Power1_3", 7),
    ("Power1_4MW", "Power1_4", 10),
    ("Power2_1MW", "Power2_1", 2),
    ("Power2_2MW", "Power2_2", 5),
    ("Power2_3MW", "Power2_3", 8),
    ("Power2_4MW", "Power2_4", 11),
    ("Power3_1MW", "Power3_1", 3),
    ("Power3_2MW", "Power3_2", 6),
]

print(f"1. Setup complete ... {time.time() - t0:.2f}s")

# ============================================================
# LOAD DC RAW DATA
# ============================================================

weekly_raw: dict[str, np.ndarray] = {}

# Source 1 — column index 8 (0-based), which is column 9 in MATLAB
for raw_name, filename in SOURCE1_FILES.items():
    df = pd.read_excel(DATA_PATH_SOURCE1 + filename)
    weekly_raw[raw_name] = df.iloc[:, 8].to_numpy(dtype=float)

# Source 2 — column index 8
for raw_name, filename in SOURCE2_FILES.items():
    df = pd.read_excel(DATA_PATH_SOURCE2 + filename)
    weekly_raw[raw_name] = df.iloc[:, 8].to_numpy(dtype=float)

# Source 3 — two sheets with different column indices
df3_1 = pd.read_excel(DATA_PATH_SOURCE3 + SOURCE3_FILE, sheet_name="Data Set 1")
weekly_raw["Power3_1MW"] = df3_1.iloc[:, 10].to_numpy(dtype=float)  # column 11 in MATLAB

df3_2 = pd.read_excel(DATA_PATH_SOURCE3 + SOURCE3_FILE, sheet_name="Data Set 2")
weekly_raw["Power3_2MW"] = df3_2.iloc[:, 1].to_numpy(dtype=float)   # column 2 in MATLAB

print(f"2. Data loading (Raw) complete ... {time.time() - t0:.2f}s")

# ============================================================
# FORMAT AND TRIM DATA
# ============================================================

MINUTES_PER_DAY = 1440
MINUTES_PER_WEEK = 10080

PERCENTILES = [0, 10, 25, 50, 75, 90, 100]
PCTILE_KEYS = ["p0", "p10", "p25", "p50", "p75", "p90", "p100"]


def compute_percentiles(power: np.ndarray, period: int) -> dict[str, np.ndarray]:
    """Reshape *power* into columns of *period* rows and compute row-wise percentiles."""
    n = (len(power) // period) * period
    mat = power[:n].reshape(period, -1, order="F")  # (period, n_periods)
    return {key: np.percentile(mat, pct, axis=1) for key, pct in zip(PCTILE_KEYS, PERCENTILES)}


daily: dict[str, dict] = {}
weekly: dict[str, dict] = {}

for raw_name, proc_name, _ in DATASETS:
    power = weekly_raw[raw_name]
    daily[proc_name] = compute_percentiles(power, MINUTES_PER_DAY)
    weekly[proc_name] = compute_percentiles(power, MINUTES_PER_WEEK)

# Time axes
t_daily = np.arange(MINUTES_PER_DAY) / 60          # hours  0–24
t_weekly = np.arange(MINUTES_PER_WEEK) / MINUTES_PER_DAY  # days  0–7

print(f"3. Data formatting and trimming complete ... {time.time() - t0:.2f}s")

# ============================================================
# HELPER: PLOT PERCENTILE PANEL
# ============================================================

FILL_COLORS = {
    "band_0_100": (0.90, 0.90, 0.90),
    "band_10_90": (0.75, 0.85, 0.95),
    "band_25_75": (0.50, 0.70, 0.90),
}


def plot_percentile_panel(ax: plt.Axes, t: np.ndarray, S: dict) -> list:
    """Fill percentile bands and plot the median on *ax*.

    Returns handles in legend order: [median, 25-75, 10-90, 0-100].
    """
    h1 = ax.fill_between(t, S["p0"],  S["p100"], color=FILL_COLORS["band_0_100"],  label="0–100%")
    h2 = ax.fill_between(t, S["p10"], S["p90"],  color=FILL_COLORS["band_10_90"],  label="10–90%")
    h3 = ax.fill_between(t, S["p25"], S["p75"],  color=FILL_COLORS["band_25_75"],  label="25–75%")
    (h4,) = ax.plot(t, S["p50"], color="blue", linewidth=1, label="50%")
    return [h4, h3, h2, h1]  # legend order matches MATLAB


# ============================================================
# FIGURE 1 — DAILY PROFILES
# ============================================================

fig1, axes1 = plt.subplots(4, 3, figsize=(18, 12), constrained_layout=True)
fig1.suptitle("Daily Power Profiles", fontsize=14)

legend_handles = None

for raw_name, proc_name, tile_id in DATASETS:
    # Convert 1-based tile_id to 0-based (row, col) for a 4×3 grid
    row, col = divmod(tile_id - 1, 3)
    ax = axes1[row, col]
    ax.grid(True)

    S = daily[proc_name]
    handles = plot_percentile_panel(ax, t_daily, S)

    if legend_handles is None:
        legend_handles = handles

    ax.set_xlabel("Hour of Day")
    ax.set_ylabel("Power (MW)")
    ax.set_title(proc_name)
    ax.set_xticks(range(0, 25))

# Hide any unused subplots (tile positions not covered by datasets)
used_tiles = {tile_id for _, _, tile_id in DATASETS}
for tile_id in range(1, 13):
    if tile_id not in used_tiles:
        row, col = divmod(tile_id - 1, 3)
        axes1[row, col].set_visible(False)

fig1.legend(
    legend_handles,
    ["50%", "25–75%", "10–90%", "0–100%"],
    loc="lower center",
    ncol=4,
    bbox_to_anchor=(0.5, -0.02),
)

# ============================================================
# FIGURE 2 — WEEKLY PROFILES
# ============================================================

fig2, axes2 = plt.subplots(4, 3, figsize=(18, 12), constrained_layout=True)
fig2.suptitle("Weekly Power Profiles", fontsize=14)

legend_handles = None

for raw_name, proc_name, tile_id in DATASETS:
    row, col = divmod(tile_id - 1, 3)
    ax = axes2[row, col]
    ax.grid(True)

    S = weekly[proc_name]
    handles = plot_percentile_panel(ax, t_weekly, S)

    if legend_handles is None:
        legend_handles = handles

    ax.set_xlabel("Day of Week")
    ax.set_ylabel("Power (MW)")
    ax.set_title(proc_name)
    ax.set_xticks(range(0, 8))
    ax.set_xticklabels(["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun", "Mon"])

for tile_id in range(1, 13):
    if tile_id not in used_tiles:
        row, col = divmod(tile_id - 1, 3)
        axes2[row, col].set_visible(False)

fig2.legend(
    legend_handles,
    ["50%", "25–75%", "10–90%", "0–100%"],
    loc="lower center",
    ncol=4,
    bbox_to_anchor=(0.5, -0.02),
)

print(f"4. Plots complete ... {time.time() - t0:.2f}s")

plt.show()

File: examples/33_data_center/process_dipolee_data.m
%%
% SETUP
% ------------------------------------------------------------
tic;
close all;
clear;
clc;
format bank;
restoredefaultpath;

% supress warnings  
warning('off','all');

% Process calibrations
DataPathname_Source1 = 'C:\Users\v0y\OneDrive - Oak Ridge National Laboratory\3. Projects\4.6 HFTO - misc\4. HFTO - SI - DC\Phase 1\Task 2 – DC Load\Samples\NLR\03_whole-facility_profiles\colocation\simulated_data\';
DataFilename_Source1_1 = 'colocation_10MW_2840nodes_20u_power.xlsx';
DataFilename_Source1_2 = 'colocation_10MW_2840nodes_40u_power.xlsx';
DataFilename_Source1_3 = 'colocation_10MW_2840nodes_60u_power.xlsx';
DataFilename_Source1_4 = 'colocation_10MW_2840nodes_80u_power.xlsx';

DataPathname_Source2 = 'C:\Users\v0y\OneDrive - Oak Ridge National Laboratory\3. Projects\4.6 HFTO - misc\4. HFTO - SI - DC\Phase 1\Task 2 – DC Load\Samples\NLR\03_whole-facility_profiles\inference\simulated_data\';
DataFilename_Source2_1 = 'inference_1MW_283nodes_20u_power.xlsx';
DataFilename_Source2_2 = 'inference_1MW_283nodes_40u_power.xlsx';
DataFilename_Source2_3 = 'inference_1MW_283nodes_60u_power.xlsx';
DataFilename_Source2_4 = 'inference_1MW_283nodes_80u_power.xlsx';

DataPathname_Source3 = 'C:\Users\v0y\OneDrive - Oak Ridge National Laboratory\3. Projects\4.6 HFTO - misc\4. HFTO - SI - DC\Phase 1\Task 2 – DC Load\Samples\ORNL\';
DataFilename_Source3 = '87F42G2C+2QG3-124-131-124-131.xlsx';

datasets = {
    'Power1_1MW', 'Power1_1', 1;
    'Power1_2MW', 'Power1_2', 4;
    'Power1_3MW', 'Power1_3', 7;
    'Power1_4MW', 'Power1_4', 10;
    'Power2_1MW', 'Power2_1', 2;
    'Power2_2MW', 'Power2_2', 5;
    'Power2_3MW', 'Power2_3', 8;
    'Power2_4MW', 'Power2_4', 11;
    'Power3_1MW', 'Power3_1', 3;
    'Power3_2MW', 'Power3_2', 6;
};

disp(['1. Setup complete ...' num2str(toc)]);

%% 
% LOAD DC RAW DATA
% ------------------------------------------------------------

Data_DC.All1_1 = readtable(strcat(DataPathname_Source1,DataFilename_Source1_1));
Data_DC.Weekly.Power1_1MW = table2array(Data_DC.All1_1(:,9));
Data_DC.All1_2 = readtable(strcat(DataPathname_Source1,DataFilename_Source1_2));
Data_DC.Weekly.Power1_2MW = table2array(Data_DC.All1_2(:,9));
Data_DC.All1_3 = readtable(strcat(DataPathname_Source1,DataFilename_Source1_3));
Data_DC.Weekly.Power1_3MW = table2array(Data_DC.All1_3(:,9));
Data_DC.All1_4 = readtable(strcat(DataPathname_Source1,DataFilename_Source1_4));
Data_DC.Weekly.Power1_4MW = table2array(Data_DC.All1_4(:,9));

Data_DC.All2_1 = readtable(strcat(DataPathname_Source2,DataFilename_Source2_1));
Data_DC.Weekly.Power2_1MW = table2array(Data_DC.All2_1(:,9));
Data_DC.All2_2 = readtable(strcat(DataPathname_Source2,DataFilename_Source2_2));
Data_DC.Weekly.Power2_2MW = table2array(Data_DC.All2_2(:,9));
Data_DC.All2_3 = readtable(strcat(DataPathname_Source2,DataFilename_Source2_3));
Data_DC.Weekly.Power2_3MW = table2array(Data_DC.All2_3(:,9));
Data_DC.All2_4 = readtable(strcat(DataPathname_Source2,DataFilename_Source2_4));
Data_DC.Weekly.Power2_4MW = table2array(Data_DC.All2_4(:,9));

Data_DC.All3_1 = readtable(strcat(DataPathname_Source3,DataFilename_Source3),'Sheet','Data Set 1');
Data_DC.Weekly.Power3_1MW = table2array(Data_DC.All3_1(:,11));
Data_DC.All3_2 = readtable(strcat(DataPathname_Source3,DataFilename_Source3),'Sheet','Data Set 2');
Data_DC.Weekly.Power3_2MW = table2array(Data_DC.All3_2(:,2));

% Data_DC.Time_mins = (1 : length(Data_DC.1_1Power_MW))';

disp(['2. Data loading (Raw) complete ...' num2str(toc)]); 

%% 
% FORMAT AND TRIM DATA
% ------------------------------------------------------------

minutes_per_day = 1440;
minutes_per_week = 10080;

for k = 1:size(datasets,1)

    raw_name = datasets{k,1};
    proc_name = char(datasets{k,2});

    power = Data_DC.Weekly.(raw_name);

    % --- reshape ---
    nD = floor(length(power)/minutes_per_day);
    nW = floor(length(power)/minutes_per_week);

    p_day = reshape(power(1:nD*minutes_per_day), minutes_per_day, nD);
    p_week = reshape(power(1:nW*minutes_per_week), minutes_per_week, nW);

    % --- compute ---
    Data_DC.Daily.(proc_name)  = compute_percentiles(p_day);
    Data_DC.Weekly.(proc_name) = compute_percentiles(p_week);

end

% time axis
t_weekly = (0:minutes_per_week-1) / 1440; % days (0–7)
t_daily = (0:minutes_per_day-1) / 60; % hours (0-23)

disp(['3. Data formating and trimming complete...' num2str(toc)]); 

%% 
% PLOT WITH SHARED BANDS
% ------------------------------------------------------------
close all;

%-------------- Figure 1 ---------------
fig = figure(1); clf; fig.WindowState = 'maximized';
tl = tiledlayout(4,3,'TileSpacing','compact','Padding','compact');

for k = 1:size(datasets,1)

    proc_name = datasets{k,2};
    tile_id   = datasets{k,3};

    nexttile(tile_id); hold on; grid on;

    S = Data_DC.Daily.(proc_name);

    H = plot_percentile_panel(t_daily, S);

    if k == 1
        H_legend = H;
    end

    xlabel('Hour of Day');
    ylabel('Power (MW)');
    title(proc_name);

    xticks(0:1:24);
end

lgd = legend(H_legend, {'50%','25–75%','10–90%','0–100%'}, 'Orientation','horizontal');
lgd.Layout.Tile = 'south';

%-------------- Figure 2 ---------------
fig = figure(2); clf; fig.WindowState = 'maximized';
tl = tiledlayout(4,3,'TileSpacing','compact','Padding','compact');

for k = 1:size(datasets,1)

    proc_name = datasets{k,2};
    tile_id   = datasets{k,3};

    nexttile(tile_id); hold on; grid on;

    S = Data_DC.Weekly.(proc_name);

    H = plot_percentile_panel(t_weekly, S);

    if k == 1
        H_legend = H;
    end

    xlabel('Day of Week');
    ylabel('Power (MW)');
    title(proc_name);

    xticks(0:1:7);
    xticklabels({'Mon','Tue','Wed','Thu','Fri','Sat','Sun','Mon'});
end

lgd = legend(H_legend, {'50%','25–75%','10–90%','0–100%'}, 'Orientation','horizontal');
lgd.Layout.Tile = 'south';


disp(['4. Plots complete ...' num2str(toc)]);

%% 
% Helper function
% 
function S = compute_percentiles(X)
    S.p50  = prctile(X,50,2);
    S.p25  = prctile(X,25,2);
    S.p75  = prctile(X,75,2);
    S.p10  = prctile(X,10,2);
    S.p90  = prctile(X,90,2);
    S.p0   = prctile(X,0,2);
    S.p100 = prctile(X,100,2);
end

function H = plot_percentile_panel(t, S)

    h1 = fill([t fliplr(t)], [S.p0' fliplr(S.p100')], ...
        [0.9 0.9 0.9], 'EdgeColor','none');

    h2 = fill([t fliplr(t)], [S.p10' fliplr(S.p90')], ...
        [0.75 0.85 0.95], 'EdgeColor','none');

    h3 = fill([t fliplr(t)], [S.p25' fliplr(S.p75')], ...
        [0.5 0.7 0.9], 'EdgeColor','none');

    h4 = plot(t, S.p50, 'b', 'LineWidth', 1);

    H = [h4 h3 h2 h1]; % legend order
end